core/ordering/cosipbft/pbft/mod.go

```diff
@@ -186,33 +188,55 @@ func (m *pbftsm) GetViews() map[mino.Address]View {
 	return views
 }
 
+// GetCommit implements pbft.StateMachine. It returns the proposal identifier
+// and the block that have been proposed to the state machine. The values are
+// valid only if the state is at least PrepareState.
+func (m *pbftsm) GetCommit() (types.Digest, types.Block) {
+	m.Lock()
+	defer m.Unlock()
+
+	return m.round.id, m.round.block
+}
+
 // Prepare implements pbft.StateMachine. It receives the proposal from the
 // leader and the current tree, and produces the next tree alongside the ID of
 // the proposal that will be signed.
-func (m *pbftsm) Prepare(block types.Block) (types.Digest, error) {
+func (m *pbftsm) Prepare(from mino.Address, block types.Block) (types.Digest, error) {
 	m.Lock()
 	defer m.Unlock()
 
-	if m.state == PrepareState {
-		// It only accepts one proposal from the leader and skip any more
-		// arriving.
-		return m.round.id, nil
-	}
+	id := m.round.id
 
-	if m.state != InitialState && m.state != NoneState {
-		return types.Digest{}, xerrors.Errorf("mismatch state %v != %v", m.state, InitialState)
+	if m.state == ViewChangeState {
+		// When in view change mode, it must refused any proposal incoming until
+		// the node leaves the state.
+		return id, xerrors.Errorf("mismatch state %v != %v", m.state, InitialState)
 	}
 
 	roster, err := m.authReader(m.tree.Get())
 	if err != nil {
-		return types.Digest{}, xerrors.Errorf("failed to read roster: %v", err)
+		return id, xerrors.Errorf("failed to read roster: %v", err)
+	}
+
+	_, index := roster.GetPublicKey(from)
+
+	if uint16(index) != m.round.leader {
+		return id, xerrors.Errorf("'%v' is not the leader", from)
+	}
+
+	// Check the state after verifying that the proposal comes from the right
+	// leader.
+	if m.state == PrepareState || m.state == CommitState {
+		// The leader should only propose one block, therefore the accepted
+		// proposal identifier is sent back, whatever the input is.
+		return id, nil
```

Comment:
then what happens to the block? The leader will wait for this block to be committed, which will reach a timeout?

Reply:
Yes, if the leader decides to propose a different block, the followers will simply reject it and proceed to a view change after the timeout. Then new leader == new block.